### DISCUSSION TITLE
Two scriptures that can help you

### DISCUSSION THREAD

**Discussion by tocacyy** [lmusarella/Pancakeswap-Prediction-Bot-2023#32] [General]
To avoid checking the console every 5 minutes to see the earnings, I decided to create this script with the help of artificial intelligence. The script, which runs in Python, is designed to send notifications on Telegram hourly with the results provided by the bot. To install and run it, you need Python installed on your system. I believe these scripts can be integrated into the bot, allowing the automation of checking and transferring funds, making the process more efficient and saving time.

import json
import time
import requests
import schedule
import os

TOKEN = os.getenv('TELEGRAM_BOT_TOKEN')
CHAT_ID = os.getenv('TELEGRAM_CHAT_ID')

if not TOKEN or not CHAT_ID:
    print("Ensure that the environment variables 'TELEGRAM_BOT_TOKEN' and 'TELEGRAM_CHAT_ID' are set correctly.")
    exit()

def send_message_to_telegram(text):
    try:
        url = f"https://api.telegram.org/bot{TOKEN}/sendMessage"
        params = {'chat_id': CHAT_ID, 'text': text}  # Here we use params instead of data
        response = requests.get(url, params=params)  # And we use the GET method instead of POST
        response.raise_for_status()

        print(f"Message sent successfully: {text}")

    except Exception as e:
        print(f"Error sending the message: {e}")

def send_statistics():
    try:
        # Replace here with the full or relative path to your file
        file_path = '/root/................................'
        with open(file_path, 'r') as file:
            data = json.load(file)

        message = f"""
Profit USD: {data['profit_usd']}
Profit Crypto: {data['profit_crypto']}
Percentage: {data['percentage']}%
Win: {data['win']}
Loss: {data['loss']}
Bet Errors: {data['betErrors']}
Total Tx Gas Fee: {data['totalTxGasFee']}
Total Tx Gas Fee USD: {data['totalTxGasFeeUsd']}
        """

        send_message_to_telegram(message)

    except Exception as e:
        print(f"Could not send statistics: {e}")

send_message_to_telegram("This is a test message.")
schedule.every(65).minutes.do(send_statistics)

while True:
    schedule.run_pending()
    time.sleep(1)

The second script is one that checks my wallet every minute, and if the balance reaches the threshold I have set, of +10%, it automatically transfers it to another wallet as profit. This way, the initial amount with which I started betting remains in the wallet I use on PancakeSwap.

from web3 import Web3
import time
import logging
import os

# Logging setup
logging.basicConfig(level=logging.INFO)

# Connect to BSC node
w3 = Web3(Web3.HTTPProvider('https://bsc-dataseed1.binance.org:443'))

# Recipient wallet address and validation
recipient_wallet_address = os.getenv('RECIPIENT_WALLET_ADDRESS')
if recipient_wallet_address is None or not w3.is_address(recipient_wallet_address):
    logging.error("The recipient wallet address is either invalid or not set!")
    exit()

# Minimum amount and percentage to be transferred
minimum_balance = 0.8  # in BNB
transfer_percentage = 10  # 10% of balance

# Time interval (in seconds) for checking the balance
check_interval = 60  # once per minute

def check_and_transfer(wallet_address, private_key):
    try:
        if wallet_address is None or not w3.is_address(wallet_address):
            logging.error(f"The wallet address {wallet_address} is either invalid or not set!")
            return

        # Check balance
        balance = w3.eth.get_balance(wallet_address)
        balance_in_bnb = w3.from_wei(balance, 'ether')

        logging.info(f"Current balance for {wallet_address}: {balance_in_bnb} BNB")

        if balance_in_bnb >= minimum_balance:
            # Calculate the amount to transfer
            transfer_amount = int(balance * transfer_percentage / 100)

            # Build the transaction
            transaction = {
                'to': recipient_wallet_address,
                'value': transfer_amount,
                'gas': 2000000,
                'gasPrice': w3.to_wei('5', 'gwei'),
                'nonce': w3.eth.get_transaction_count(wallet_address),
            }

            # Sign the transaction
            signed_transaction = w3.eth.account.sign_transaction(transaction, private_key)

            # Send the transaction and check the status
            transaction_hash = w3.eth.send_raw_transaction(signed_transaction.rawTransaction)
            logging.info(f"Transaction sent with hash: {transaction_hash.hex()}")

    except Exception as e:
        logging.error(f"Error in transaction for {wallet_address}: {e}")

while True:
    wallets = {
        os.getenv('WALLET_ADDRESS_1'): os.getenv('PRIVATE_KEY_1'),
        #os.getenv('WALLET_ADDRESS_2'): os.getenv('PRIVATE_KEY_2'),
    }

    if None in wallets.keys() or None in wallets.values():
        logging.error("Wallet addresses or private keys are not properly set in the environment variables.")
        exit()

    for wallet_address, private_key in wallets.items():
        check_and_transfer(wallet_address, private_key)
    time.sleep(check_interval)

**Comment by jtruji68** [comment]
Hi, thanks, Ill use the telegram one. You seem like a dev so Ill ask: I was looking at the current code and I could not find where to setup the gas Price, I just found a call to an event 
`  const tx = await getSmartContract().betBull(epoch, getAmountValue(amount));`
but I cant see where the hell is the transaction body. I just found the event that is called after the transaction is made:
```
getSmartContract().on(EVENTS.BET_BULL_EVENT, async (sender, epoch, betAmount) => {
  const round = formatUnit(epoch);

  //Check if the bot has to register the user activity for this round
  if(GLOBAL_CONFIG.ANALYTICS_CONFIGURATION.REGISTER_USERS_ACTIVITY) {
    registerUser(round, sender, BET_UP, formatUnit(betAmount, "18"));
  } 
```
Do you know where i can find it? from what you have seen in the code.
Thanks

**Reply by tocacyy** [reply]
From what I've noticed, considering my programming knowledge, I'm a beginner too :), this is done automatically in the wallet.module.js. This is the function, in my opinion, but I could be wrong:
/**

Get your gas price of BSC chain
@async
@returns {number} gas price
*/
const getGasPrice = async () => {
return formatUnit(await provider.getGasPrice());

I also had this problem and I reduced it by changing: BSC_NODE_PROVIDER_URL='https://bscrpc.com/' to about 600 rounds I had two errors.
Please consider that I'm not that good at programming; I'm a beginner in this. The only person who can confirm if it is as I say is @lmusarella :)

**Reply by jtruji68** [reply]
Thanks @tocacyy , your provider_url worked pretty well, didnt have to mess with the code anymore. Also your telegram idea was good. I already had a bot so setting it up was easy.
